--- a/glance/common/auth.py
+++ b/glance/common/auth.py
@@ -148,8 +148,15 @@
     def _v2_auth(self, token_url):
         creds = self.creds
 
-        creds = {"passwordCredentials": {"username": creds['username'],
-                                         "password": creds['password']}}
+        creds = {
+            "auth": {
+                "tenantName": creds['tenant'],
+                "passwordCredentials": {
+                    "username": creds['username'],
+                    "password": creds['password']
+                    }
+                }
+            }
 
         tenant = creds.get('tenant')
         if tenant:
@@ -163,12 +170,16 @@
                 token_url, 'POST', headers=headers, body=req_body)
 
         if resp.status == 200:
-            resp_auth = json.loads(resp_body)['auth']
+            resp_auth = json.loads(resp_body)['access']
 
             # FIXME(sirp): for now just using the first endpoint we get back
             # from the service catalog for glance, and using the public url.
-            glance_info = resp_auth['serviceCatalog']['glance']
-            glance_endpoint = glance_info[0]['publicURL']
+            for service in resp_auth['serviceCatalog']:
+                if service['name'] == 'glance':
+                    glance_endpoint = service['endpoints'][0]['publicURL']
+                    break
+            else:
+                raise exception.NoServiceEndpoint()
 
             self.management_url = glance_endpoint
             self.auth_token = resp_auth['token']['id']
--- a/glance/common/exception.py
+++ b/glance/common/exception.py
@@ -172,3 +172,7 @@
 
 class InvalidNotifierStrategy(GlanceException):
     message = "'%(strategy)s' is not an available notifier strategy."
+
+
+class NoServiceEndpoint(GlanceException):
+    message = _("Response from Keystone does not contain a Glance endpoint.")
